sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala

```diff
@@ -948,6 +952,210 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
+      inputTypes.map(generateSingleEntry(_, collationType))
+      }
+
+    val toSkip = List(
+      "get_json_object",
+      "map_zip_with",
+      "printf",
+      "transform_keys",
+      "concat_ws",
+      "format_string",
+      "session_window",
+      "transform_values",
+      "arrays_zip",
+      "hex" // this is fine
+    )
+
+    for (f <- funInfos.filter(f => !toSkip.contains(f.getName))) {
+      val cl = Utils.classForName(f.getClassName)
+      val headConstructor = cl.getConstructors.head
+      val params = headConstructor.getParameters.map(p => p.getType)
+      val paramCount = params.length
+      val args = params.map {
+        case e if e.isAssignableFrom(classOf[Expression]) => Literal.create("1")
+        case se if se.isAssignableFrom(classOf[Seq[Expression]]) =>
+          Seq(Literal.create("1"))
+        case oe if oe.isAssignableFrom(classOf[Option[Expression]]) => None
+      }
+      val expr = headConstructor.newInstance(args: _*).asInstanceOf[ExpectsInputTypes]
+      val inputTypes = expr.inputTypes
+
+      var nones = Array.fill(0)(None)
+      if (paramCount > inputTypes.length) {
+        nones = Array.fill(paramCount - inputTypes.length)(None)
+      }
+
+      val inputDataUtf8Binary = generateData(inputTypes.take(paramCount), Utf8Binary) ++ nones
+      val instanceUtf8Binary =
+        headConstructor.newInstance(inputDataUtf8Binary: _*).asInstanceOf[Expression]
+
+      val inputDataLcase = generateData(inputTypes.take(paramCount), Utf8BinaryLcase) ++ nones
+      val instanceLcase = headConstructor.newInstance(inputDataLcase: _*).asInstanceOf[Expression]
+
+      val exceptionUtfBinary = {
+        try {
+          instanceUtf8Binary.eval(EmptyRow)
+          None
+        } catch {
+          case e: Throwable => Some(e)
+        }
+      }
+
+      val exceptionLcase = {
+        try {
+          instanceLcase.eval(EmptyRow)
+          None
+        } catch {
+          case e: Throwable => Some(e)
+        }
+      }
+
+      // Check that both cases either throw or pass
+      assert(exceptionUtfBinary.isDefined == exceptionLcase.isDefined)
+
+      if (exceptionUtfBinary.isEmpty) {
+        val resUtf8Binary = instanceUtf8Binary.eval(EmptyRow)
+        val resUtf8Lcase = instanceLcase.eval(EmptyRow)
+
+        val dt = instanceLcase.dataType
+
+        dt match {
+          case _: StringType if resUtf8Lcase != null && resUtf8Lcase != null =>
+            assert(resUtf8Binary.isInstanceOf[UTF8String])
+            assert(resUtf8Lcase.isInstanceOf[UTF8String])
+            // scalastyle:off caselocale
+            assert(resUtf8Binary.asInstanceOf[UTF8String].toLowerCase.binaryEquals(
+              resUtf8Lcase.asInstanceOf[UTF8String].toLowerCase))
+            // scalastyle:on caselocale
+          case _ => resUtf8Lcase === resUtf8Binary
+        }
+      }
+      else {
+        assert(exceptionUtfBinary.get.getClass == exceptionLcase.get.getClass)
```

Comment:
Can you add some way to monitor the number of covered expressions? E.g. either some test log or at least comment with code that specifies the number of covered expressions.